### ISSUE TITLE
Subpath: This route collides with: "src/pages/404.astro".

### ISSUE THREAD

**Issue by rwieruch** [withastro/starlight#1080]
### What version of `starlight` are you using?

0.12.1

### What version of `astro` are you using?

3.3.0

### What package manager are you using?

yarn

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

I already used Astro for my website and wanted to integrate Astro Starlight manually  to it for my "Help Desk". When I tried to start it, I got the following error:

```
This route collides with: "src/pages/404.astro".
```

Because Astro Starlight comes with a 404 as well, I had to remove my top-level 404 page. It works, however, are there any plans to allow two 404 if Starlight gets dropped in as a [subpath](https://starlight.astro.build/manual-setup/#tips-for-existing-sites)?

Thanks for all your work here!

### Link to Minimal Reproducible Example

-

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by rwieruch** [comment]
Thanks for the quick reply @delucis 

I think opt-ing out Starlights 404 is okay. Perhaps it would be even better to serve the Starlights 404 only if something within the subpath does not match. If soemthing is not found outside of the subpath, meaning outside Starlight, it shoudl show the top-level 404. 

I come to this conclusion because the top-level website looks different from Starlight, so it should be okay to show the 404 within each's context for aesthetic reasons.

**Comment by delucis** [comment]
> Perhaps it would be even better to serve the Starlights 404 only if something within the subpath does not match.

That’s probably outside of our control unfortunately without users setting up some custom config on their host. Most static hosts just treat a root `404.html` as the 404 page and don’t support multiple. You can work around this with custom config — we do this for multilingual 404s in Starlight’s docs for example — but we couldn’t offer it as a default “works everywhere” behaviour.

#### Example configs

##### `netlify.toml`

```toml
[[redirects]]
  from = "/:lang/*"
  to = "/:lang/404/"
  status = 404
```

##### `vercel.json`

```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "routes": [
    { "src": "/(?<lang>[^/]*)/(.*)", "dest": "/$lang/404/", "status": 404 }
  ]
}
```

(Unfortunately requires Vercel’s deprecated `routes` config from what I’ve gathered.)

**Comment by rwieruch** [comment]
Thanks for the insights and your time @delucis Very much appreciated! For my case everything works fine and I am amazed how smoothly the Starlight integration went for my existing Astro website. This was next level! Feel free to close this issue or to move it to the discussion section of GitHub :)

**Comment by delucis** [comment]
Just tested, and the same limitation around moving this to the dynamic route still exists in Astro. I’m going to talk to the team to see if this is something we can fix upstream. If it isn’t, we can look at adding a flag to disable Starlight’s 404 injection.

---

<sup>🔗 [Internal discussion reference for Astro maintainers](https://discord.com/channels/830184174198718474/1175066542510198824)</sup>

**Comment by abs** [comment]
My project [ab]uses the 404.astro page to load the SPA:

![image](https://github.com/withastro/starlight/assets/41166/f46d427a-83fc-40c4-9b56-1dc82944b1d5)

I've been looking for workarounds, but so far it's all looking pretty bleak :)

If there was an easy way to turn off Starlight's 404 injection, it would help a ton.

**Comment by delucis** [comment]
> If there was an easy way to turn off Starlight's 404 injection, it would help a ton.

We’ve been discussing it — hopefully will know more next week!

In the meantime, you _could_ add another level of hackiness by loading your SPA from `src/content/docs/404.mdx`:

```
---
title: My SPA
---

import SPA from '../../SPA.astro';

<SPA client:load />
```

**Comment by abs** [comment]
> In the meantime, you _could_ add another level of hackiness by loading your SPA from `src/content/docs/404.mdx`:

I did try that, but of course it loads the SPA from inside Starlight (instead of its 404), which is... suboptimal, to put it mildly :-)

I've unblocked our docs development by using a fork - https://github.com/abs/starlight/commit/db448f42754607a0b93e3ab07b47c719f2853fd3

Once y'all come up with a fix, I'd be happy to give it a go!

**Comment by delucis** [comment]
Ohhhh, yeah of course 😅 

Will let everyone know more here next week 🙌

**Comment by olton** [comment]
Hello. Any news on how to fix this? `This route collides with: "src/pages/404.astro"`

**Comment by delucis** [comment]
Fair question! No progress just yet, but https://github.com/withastro/astro/pull/9439 will allow us to fix this really nicely by enabling us to set a lower priority on Starlight’s injected 404 route. This would let user 404s take over if they exist.

https://github.com/withastro/astro/pull/9439 is a larger PR to Astro and we’re heading towards the holiday period, so I don’t want to promise too much with regards to the timeline, but at least a solution is on the horizon!

**Comment by delucis** [comment]
https://github.com/withastro/astro/pull/9439 ended up taking a different direction and won’t help with this, so I think it’s time to add a configuration flag to help unblock folks on this — we can always remove the flag in the future if we have better options.

Would someone like to contribute a PR? I’d suggest something like:

1. Add an option for disabling 404 rendering to the user config

2. Skip the `injectRoute()` call if that is set

3. Update configuration reference with the new option and maybe also the [“Custom 404 page”](https://starlight.astro.build/guides/customization/#custom-404-page) section in the docs

**Comment by connor-baer** [comment]
I'd be happy to take a stab at it. Thanks for the helpful instructions to get started!

**Comment by delucis** [comment]
Thanks @connor-baer! Let me know if you have any questions along the way 🙌
